### PR TITLE
Provide a mechanism to let fixtures skip tests

### PR DESCRIPTION
**Pull request by richvdh**
In some situations, it turns out to make more sense to skip tests when certain fixtures fail than to fail them. Make it so that if a fixture fails with "SKIP: ...", then we skip any tests which require it rather than failing them.

(This PR includes 3 commits which it may be useful to review individually.)

### CODE REVIEW

File: run-tests.pl [hunk 1]
```diff
@@ -525,9 +525,40 @@ sub _run_test
...
+   }
+   else {
+      $t->fail( $reason );
+   }
+}
+
+# Helper for _run_test. Waits for the req_futures to become ready, then runs
+# the test itself and waits for the test to complete.
+#
+# returns a pair [$res, $reason] where $res is one of:
```
Comment (leonerd) [inline run-tests.pl]: Best to document that with `(parens)` so the reader doesn't think it returns a two-element arrayref.

File: DEVELOP.rst [hunk 2]
```diff
@@ -200,6 +200,14 @@ main kinds of fixtures:
   over the lifetime of the test. These are fixtures that have a ``teardown``
   block.
 
+In general, if the future returned by ``setup`` fails, then any tests which
+require that fixture will be marked as failures. However, if the future fails
+with a reason beginning with the special string ``SKIP``, then any tests which
+require it are instead skipped. This can be used to differentiate between tests
+which we were unable to run due to constraints when setting up the test
+environment, as opposeed to those where setup steps that should have succeeded
```
Comment (leonerd) [inline DEVELOP.rst]: opposed

File: run-tests.pl [hunk 3]
```diff
@@ -549,17 +552,29 @@ sub _run_test
...
       my $f_setup = Future->needs_all( @$req_futures )
          ->on_done( sub { @reqs = @_ } )
-         ->on_fail( sub { die "fixture failed - $_[0]\n" } );
+         ->on_fail( sub {
+            my ( $reason ) = @_;
+
+            # if any of the fixtures failed with a special 'SKIP' result, then skip
+            # the test rather than failing it.
+            if ( $reason =~ /^SKIP(: *(.*))?$/ ) {
+               my $r = ": $2" // '';
```
Comment (leonerd) [inline run-tests.pl]: The string `": $2"` is never going to be undefined here even if `$2` was. You probably want to use ternary instead

    my $r = defined $2 ? ": $2" : "";

File: run-tests.pl [hunk 4]
```diff
@@ -549,17 +552,29 @@ sub _run_test
...
+               $skip_reason = "failing fixture$r";
```
Comment (leonerd) [inline run-tests.pl]: Or maybe:

    $skip_reason = "failing fixture";
    $skip_reason .= ": $2" if defined $2;

File: run-tests.pl (hunk 1 above)
Comment (richvdh) [reply run-tests.pl]: done in b8a39cb

File: DEVELOP.rst (hunk 2 above)
Comment (richvdh) [reply DEVELOP.rst]: done in b8a39cb

File: run-tests.pl (hunk 4 above)
Comment (richvdh) [reply run-tests.pl]: done in b8a39cb

**Comment by richvdh** [comment]
@leonerd: PTAL

**Comment by richvdh** [comment]
as per https://github.com/matrix-org/sytest/pull/374#issuecomment-330774846, merging this. The test failure seems unrelated.

**Comment by leonerd** [comment]
Retroactive LGTM

**Pull request merged**